### ISSUE TITLE
Trim the common path (eval) stack

### ISSUE THREAD

**Issue by alexander-yakushev** [nrepl/nrepl#331]
A bare no-extra-middleware stack of an evaluated expression in nREPL currently looks like this:

```
$  clojure -Sdeps '{:deps {nrepl/nrepl {:mvn/version "1.1.2"}}}' -m nrepl.cmdline -i                                                                         ~

user=> (Exception.)
#error {
 :cause nil
 :via
 [{:type java.lang.Exception
   :message nil
   :at [user$eval2608 invokeStatic "NO_SOURCE_FILE" 1]}]
 :trace
 [[user$eval2608 invokeStatic "NO_SOURCE_FILE" 1]
  [user$eval2608 invoke "NO_SOURCE_FILE" 1]
  [clojure.lang.Compiler eval "Compiler.java" 7534]
  [clojure.lang.Compiler eval "Compiler.java" 7489]
  [clojure.core$eval invokeStatic "core.clj" 3230]
  [clojure.core$eval invoke "core.clj" 3226]
  [nrepl.middleware.interruptible_eval$evaluate$fn__1477$fn__1478 invoke "interruptible_eval.clj" 87]
  [clojure.lang.AFn applyToHelper "AFn.java" 152]
  [clojure.lang.AFn applyTo "AFn.java" 144]
  [clojure.core$apply invokeStatic "core.clj" 667]
  [clojure.core$with_bindings_STAR_ invokeStatic "core.clj" 1990]
  [clojure.core$with_bindings_STAR_ doInvoke "core.clj" 1990]
  [clojure.lang.RestFn invoke "RestFn.java" 428]
  [nrepl.middleware.interruptible_eval$evaluate$fn__1477 invoke "interruptible_eval.clj" 87]
  [clojure.main$repl$read_eval_print__9216$fn__9219 invoke "main.clj" 437]
  [clojure.main$repl$read_eval_print__9216 invoke "main.clj" 437]
  [clojure.main$repl$fn__9225 invoke "main.clj" 459]
  [clojure.main$repl invokeStatic "main.clj" 459]
  [clojure.main$repl doInvoke "main.clj" 368]
  [clojure.lang.RestFn invoke "RestFn.java" 1526]
  [nrepl.middleware.interruptible_eval$evaluate invokeStatic "interruptible_eval.clj" 84]
  [nrepl.middleware.interruptible_eval$evaluate invoke "interruptible_eval.clj" 56]
  [nrepl.middleware.interruptible_eval$interruptible_eval$fn__1510$fn__1514 invoke "interruptible_eval.clj" 152]
  [clojure.lang.AFn run "AFn.java" 22]
  [nrepl.middleware.session$session_exec$main_loop__1580$fn__1584 invoke "session.clj" 218]
  [nrepl.middleware.session$session_exec$main_loop__1580 invoke "session.clj" 217]
  [clojure.lang.AFn run "AFn.java" 22]
  [java.lang.Thread run "Thread.java" 1570]]}
user=>
```

While this is no big deal, we can do a bit better to reduce the stack size. I mean doing more aggressive inlining in the most fundamental middleware like `session`, `print`, `interruptible-eval`, using fewer varargs functions, etc. Even though IDEs like CIDER hide such tooling noise from the user, reducing the stack everywhere helps in the general case. It also marginally improves debugging and profiling experience.

**Comment by cichli** [comment]
Your example illustrates the stack size when processing a request – here's one for sending a response.

```
user=> (take 1 (repeatedly (fn [] (Exception.))))
(#error {
 :cause nil
 :via
 [{:type java.lang.Exception
   :message nil
   :at [user$eval2524$fn__2525 invoke "NO_SOURCE_FILE" 1]}]
 :trace
 [[user$eval2524$fn__2525 invoke "NO_SOURCE_FILE" 1]
  [clojure.core$repeatedly$fn__6531 invoke "core.clj" 5175]
  [clojure.lang.LazySeq sval "LazySeq.java" 42]
  [clojure.lang.LazySeq seq "LazySeq.java" 51]
  [clojure.lang.RT seq "RT.java" 535]
  [clojure.core$seq__5467 invokeStatic "core.clj" 139]
  [clojure.core$take$fn__5978 invoke "core.clj" 2901]
  [clojure.lang.LazySeq sval "LazySeq.java" 42]
  [clojure.lang.LazySeq seq "LazySeq.java" 51]
  [clojure.lang.RT seq "RT.java" 535]
  [clojure.core$seq__5467 invokeStatic "core.clj" 139]
  [clojure.core$print_sequential invokeStatic "core_print.clj" 53]
  [clojure.core$fn__7391 invokeStatic "core_print.clj" 174]
  [clojure.core$fn__7391 invoke "core_print.clj" 174]
  [clojure.lang.MultiFn invoke "MultiFn.java" 234]
  [nrepl.middleware.print$pr_on invokeStatic "print.clj" 21]
  [nrepl.middleware.print$pr_on invoke "print.clj" 17]
  [nrepl.middleware.print$send_nonstreamed$print_key__1294$fn__1295 invoke "print.clj" 148]
  [nrepl.middleware.print$send_nonstreamed$print_key__1294 invoke "print.clj" 147]
  [clojure.core$map$fn__5931$fn__5932 invoke "core.clj" 2759]
  [clojure.core.protocols$iter_reduce invokeStatic "protocols.clj" 49]
  [clojure.core.protocols$fn__8230 invokeStatic "protocols.clj" 75]
  [clojure.core.protocols$fn__8230 invoke "protocols.clj" 75]
  [clojure.core.protocols$fn__8178$G__8173__8191 invoke "protocols.clj" 13]
  [clojure.core$transduce invokeStatic "core.clj" 6948]
  [clojure.core$transduce invoke "core.clj" 6934]
  [nrepl.middleware.print$send_nonstreamed invokeStatic "print.clj" 156]
  [nrepl.middleware.print$send_nonstreamed invoke "print.clj" 138]
  [nrepl.middleware.print$printing_transport$reify__1311 send "print.clj" 174]
  [nrepl.middleware.caught$caught_transport$reify__1346 send "caught.clj" 58]
  [nrepl.middleware.interruptible_eval$evaluate$fn__1412 invoke "interruptible_eval.clj" 123]
  [clojure.main$repl$read_eval_print__9206 invoke "main.clj" 442]
  [clojure.main$repl$fn__9215 invoke "main.clj" 458]
  [clojure.main$repl invokeStatic "main.clj" 458]
  [clojure.main$repl doInvoke "main.clj" 368]
  [clojure.lang.RestFn invoke "RestFn.java" 1523]
  [nrepl.middleware.interruptible_eval$evaluate invokeStatic "interruptible_eval.clj" 84]
  [nrepl.middleware.interruptible_eval$evaluate invoke "interruptible_eval.clj" 56]
  [nrepl.middleware.interruptible_eval$interruptible_eval$fn__1422$fn__1426 invoke "interruptible_eval.clj" 152]
  [clojure.lang.AFn run "AFn.java" 22]
  [nrepl.middleware.session$session_exec$main_loop__1492$fn__1496 invoke "session.clj" 218]
  [nrepl.middleware.session$session_exec$main_loop__1492 invoke "session.clj" 217]
  [clojure.lang.AFn run "AFn.java" 22]
  [java.lang.Thread run "Thread.java" 1583]]})
```

There are definitely some easy wins, e.g. 

1. Switching to `loop` instead of `run!` / `transduce` and inlining the `print-key` closure in `print/send-streamed` / `send-nonstreamed` respectively. This would trim approximately ten frames – I'll PR this if you think it's worth it.

2. Inlining `print/send-streamed` / `print/send-nonstreamed`. This would only save two frames but would greatly inflate the size of the `send` method in `print/printing-transport`.

**Comment by alexander-yakushev** [comment]
Thank you for looking into this! I have my own refactoring of `middleware.print` priming for after 1.2 is released. I'll push it a bit later so that we can iterate on it, I think. Overall, I'm not a big fun of `loop` for iterating collections since it either requires iterators (which I do use privately but it's not "the clojure way") or is inefficient. I think `reduce-kv` would do here both performance and stack-wise. But performance is honestly not important here at all. Anyway, gotta look at different options.

> Inlining print/send-streamed / print/send-nonstreamed. This would only save two frames but would greatly inflate the size of the send method in print/printing-transport.
Not worth doing, IMO. Those functions are well separated right now. What I did is I rewrote the proxies to Java classes since proxies are a mess. Also coming soon in that post-1.2 branch.

**Comment by alexander-yakushev** [comment]
> In the past I have thought about the problem we have with stack depths more generally (i.e. not just the common path) because we rely on: [function composition for handling requests]

Yes! I've been thinking about that as well. I quite don't like how unrelated middleware just stacks one upon another making the debugging more confusing. I've been thinking if we could use the current `expects/requires` data to build a tree of middleware instead of chain. Another option is to make an official metadata key that says "this middleware only handles such and such op, otherwise don't call" – similar to what cider-nrepl currently uses for deferring its middleware loading. This change can be done non-intrusively, in such a way that if the middleware is missing, then the middleware always gets chained. What would you say, @bbatsov?

**Comment by cichli** [comment]
> Overall, I'm not a big fun of loop for iterating collections since it either requires iterators (which I do use privately but it's not "the clojure way")

I think it's fine in at least the `send-streamed` case since it's entirely side-effecting anyway. For `send-nonstreamed` switching to `reduce` and inlining the `print-key` closure would maybe suffice. (N.B. I think we'd need `reduce`, not `reduce-kv`, since we're iterating over a set of keys)

> But performance is honestly not important here at all. Anyway, gotta look at different options.

Agreed.

> What I did is I rewrote the proxies to Java classes since proxies are a mess.

I did consider doing this in the first place, but can't quite recall now why I didn't. Possibly because I thought the Java->Clojure interop of calling `transport/send` would be awkward – I do agree their impls are generally better suited to Java though. One note I do still have is that if I ever rewrote this code I would switch from using `PrintWriter` to just `Writer` and deprecate `replying-PrintWriter`. I think we should just provide a `Writer` in the API and let callers wrap it in a `PrintWriter` if and only when it's needed, since `PrintWriter` swallows/hides errors.

Thanks for giving the printing code another set of eyes and a bit of a spruce up – it hasn't been touched in years.

> I've been thinking if we could use the current expects/requires data to build a tree of middleware instead of chain.

This would also have the benefit of making it more obvious if you forget to add something to expects or requires – right now, if you do that then it's a matter of luck whether or not your middleware gets called in the right order. I've fixed a missing entry in `requires` in more than one third-party middleware in the past. One big possible downside: would it make it impossible to implement a middleware that should be included unconditionally (e.g. some kind of request logging/tracing)? Not sure if there's a real/valid use case there.

> Another option is to make an official metadata key that says "this middleware only handles such and such op, otherwise don't call" – similar to what cider-nrepl currently uses for deferring its middleware loading.

I think I prefer the first option (if we can make it work) but I have been thinking about how we can support deferred loading of middleware. I'll spin that discussion out into a separate issue though, to keep this one on-topic.
